Mathlib/MeasureTheory/MeasurableSpace/CountablyGenerated.lean

```diff
@@ -0,0 +1,236 @@
+/-
+Copyright (c) 2023 Felix Weilacher. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Felix Weilacher, Yury G. Kudryashov, Rémy Degenne
+-/
+import Mathlib.MeasureTheory.MeasurableSpace.Basic
+import Mathlib.Data.Set.MemPartition
+import Mathlib.Order.Filter.CountableSeparatingOn
+
+/-!
+# Countably generated measurable spaces
+
+We say a measurable space is countably generated if it can be generated by a countable set of sets.
+
+In such a space, we can also build a sequence of finer and finer finite measurable partitions of
+the space such that the measurable space is generated by the union of all partitions.
+
+## Main definitions
+
+* `MeasurableSpace.CountablyGenerated`: class stating that a measurable space is countably
+  generated.
+* `MeasurableSpace.countablePartition`: sequences of finer and finer partitions of
+  a countably generated space.
+* `MeasurableSpace.partitionSet`: `partitionSet n t` is the set in the partition
+  `countablePartition α n` to which `t` belongs.
+
+## Main statements
+
+* `MeasurableSpace.measurable_injection_nat_bool_of_countablyGenerated`: if a measurable space is
+  countably generated and separates points, it admits a measurable injection into the Cantor space
+  `ℕ → Bool` (equipped with the product sigma algebra).
+
+-/
+
+
+open Set MeasureTheory
+
+namespace MeasurableSpace
+
+variable {α β : Type*}
+
+/-- We say a measurable space is countably generated
+if it can be generated by a countable set of sets. -/
+class CountablyGenerated (α : Type*) [m : MeasurableSpace α] : Prop where
+  isCountablyGenerated : ∃ b : Set (Set α), b.Countable ∧ m = generateFrom b
+#align measurable_space.countably_generated MeasurableSpace.CountablyGenerated
+
+/-- A countable set of sets that generate the measurable space.
+We insert `∅` to ensure it is nonempty. -/
+def generatingSet (α : Type*) [MeasurableSpace α] [h : CountablyGenerated α] : Set (Set α) :=
```

Comment:
This name is a little bit too generic. `countableGeneratingSet` would probably be better, as the point is really the countability.